pkgs/native_assets_cli/lib/test.dart

```diff
@@ -27,50 +34,72 @@ Future<void> testBuildHook({
-        )
-        ..setupBuildConfig(
-          dryRun: false,
-          linkingEnabled: true,
-        )
-        ..setupBuildRunConfig(
-          outputDirectory: outputDirectory,
-          outputDirectoryShared: outputDirectoryShared,
-        );
-      extraConfigSetup(configBuilder);
-
-      final config = BuildConfig(configBuilder.json);
-
-      final configUri = tempDir.resolve(Hook.build.outputName);
-      _writeJsonTo(configUri, config.json);
-      await mainMethod(['--config=${configUri.toFilePath()}']);
-      final output = BuildOutput(
-          _readJsonFrom(config.outputDirectory.resolve(Hook.build.outputName)));
-
-      // Test conformance of protocol invariants.
-      expect(await validateBuildOutput(config, output), isEmpty);
-
-      // Run user-defined tests.
-      check(config, output);
-    },
-  );
+  const keepTempKey = 'KEEP_TEMPORARY_DIRECTORIES';
+
+  final tempDir = await Directory.systemTemp.createTemp();
+
+  try {
+    // Deal with Windows temp folder aliases.
+    final tempUri =
+        Directory(await tempDir.resolveSymbolicLinks()).uri.normalizePath();
+    final outputDirectory = tempUri.resolve('output/');
+    final outputDirectoryShared = tempUri.resolve('output_shared/');
+
+    await Directory.fromUri(outputDirectory).create();
+    await Directory.fromUri(outputDirectoryShared).create();
+
+    final configBuilder = BuildConfigBuilder();
+    configBuilder
+      ..setupHookConfig(
+        packageRoot: Directory.current.uri,
+        packageName: _readPackageNameFromPubspec(),
+        targetOS: targetOS ?? OS.current,
+        buildAssetTypes: buildAssetTypes ?? [],
+        buildMode: buildMode ?? BuildMode.release,
+      )
+      ..setupBuildConfig(
+        dryRun: false,
+        linkingEnabled: true,
+      )
+      ..setupBuildRunConfig(
+        outputDirectory: outputDirectory,
+        outputDirectoryShared: outputDirectoryShared,
+      );
+    extraConfigSetup(configBuilder);
+
+    final config = BuildConfig(configBuilder.json);
+
+    final configUri = tempUri.resolve(Hook.build.outputName);
+    _writeJsonTo(configUri, config.json);
+    await mainMethod(['--config=${configUri.toFilePath()}']);
+    final output = BuildOutput(
+        _readJsonFrom(config.outputDirectory.resolve(Hook.build.outputName)));
+
+    // Test conformance of protocol invariants.
+    final validationErrors = await validateBuildOutput(config, output);
+    if (validationErrors.isNotEmpty) {
+      throw VerificationException(
+          'encountered build output validation issues: $validationErrors');
+    }
+
+    // Run user-defined tests.
+    await check(config, output);
+  } finally {
+    final keepTempDir = (Platform.environment[keepTempKey] ?? '').isNotEmpty;
+    if (!keepTempDir) {
+      tempDir.deleteSync(recursive: true);
+    }
+  }
+}
+
+/// An exception thrown when build hook verification fails.
+class VerificationException implements Exception {
```

Comment:
I don't think this class should be part of the public API. I don't believe users should catch it. So maybe it should an `Error` instead? `Error` itself has a message, so we could even throw `Error` instead of `_VerificationError`.

Or do you believe users should catch this?

Reply:
OK, I looked into this a bit more. `Error` itself doesn't have a message; it's subclasses do. None of those look like great candidates to throw, and, it looks like `expect` throws an Exception, not an Error, so I think we should do the same.

I updated this to be a `ValidationFailure`, implementing an Exception and not an error. I also moved it to a place in the code where it won't be public API.